client/my-sites/checkout/cart/cart-coupon.jsx

```diff
@@ -37,87 +34,125 @@ class CartCoupon extends React.Component {
 		}
 	}
 
-	toggleCouponDetails = event => {
-		event.preventDefault();
-
-		this.setState( { isCouponFormShowing: ! this.state.isCouponFormShowing } );
-
-		if ( this.state.isCouponFormShowing ) {
-			analytics.ga.recordEvent( 'Upgrades', 'Clicked Hide Coupon Code Link' );
-		} else {
-			analytics.ga.recordEvent( 'Upgrades', 'Clicked Show Coupon Code Link' );
+	render() {
+		if ( this.appliedCouponCode ) {
+			return this.renderAppliedCoupon();
 		}
-	};
 
-	applyCoupon = event => {
-		event.preventDefault();
-
-		analytics.tracks.recordEvent( 'calypso_checkout_coupon_submit', {
-			coupon_code: this.state.couponInputValue,
-		} );
+		return this.renderApplyCouponUI();
+	}
 
-		this.setState( {
-			userChangedCoupon: false,
-			hasSubmittedCoupon: true,
-		} );
-		applyCoupon( this.state.couponInputValue );
-	};
+	get appliedCouponCode() {
+		return this.props.cart.is_coupon_applied && this.props.cart.coupon
+			? this.props.cart.coupon
+			: null;
+	}
 
-	handleCouponInput = event => {
-		this.setState( {
-			userChangedCoupon: true,
-			couponInputValue: event.target.value,
-		} );
-	};
+	renderAppliedCoupon() {
+		return (
+			<div className="cart__coupon">
+				<span className="cart__details">
+					{ this.props.translate( 'Coupon applied: %(coupon)s', {
+						args: { coupon: this.appliedCouponCode },
+					} ) }
+				</span>{' '}
+				<button onClick={ this.clearCoupon } className="button is-link cart__remove-link">
+					{ this.props.translate( 'Remove' ) }
+				</button>
+			</div>
+		);
+	}
 
-	getToggleLink = () => {
+	renderApplyCouponUI() {
 		if ( this.props.cart.total_cost === 0 ) {
 			return;
```

Comment:
This is unsafe to render and likely caused errors under certain circumstanced. Should be render `null`. Could optionally be merged with the `return` below as `return this.props.cart.total_cost === 0 && ( …`

Props to @ockham for spotting.